friend_code.py

```diff
@@ -0,0 +1,86 @@
+import re
+import sqlite3
+
+import config
+
+if config.bot.acnh_db is not None:
+	db = sqlite3.connect(config.bot.acnh_db)
+	db.row_factory = sqlite3.Row
+
+	# enable foreign key constraints
+	with db:
+		db.execute('PRAGMA foreign_keys = ON')
+
+friend_code_regex = re.compile(r'SW-\d{4}-\d{4}-\d{4}')
+
+def friend_code(cmd):
+	if not cmd.args:
+		_user_list_all(cmd)
+		return
+
+	split = cmd.args.split(' ', 1)
+	subcmd = split[0]
+	if subcmd == 'set':
+		_user_upsert_friend_code(cmd, split)
+	elif subcmd == 'remove':
+		_user_remove(cmd)
+	else:
+		_user_find(cmd, split[0])
+
+def _user_upsert_friend_code(cmd, split):
+	if len(split) != 2:
+		cmd.reply('usage: !fc set friend-code')
+		return
+
+	if friend_code_regex.match(split[1]) is None:
+		cmd.reply('Invalid friend code submitted.')
+		return
+
+	sender = cmd.sender
+	cur = None
+	with db:
+		cur = db.execute('''
+		INSERT INTO user VALUES(?, ?, ?, null)
+		ON CONFLICT(id)
+		DO UPDATE SET code=excluded.code
+		''', (sender['id'], sender['username'], split[1]))
```

Comment:
ok so it turns out usernames are changeable. I've been avoiding adding ID->name tracking to the bot but it feels like it'd be pretty useful here... then we could just store the discord snowflake ID and use the bot's ID->name tracker for display